ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp

```diff
@@ -284,10 +296,56 @@ static int HandleVnodeOperation(
                 }
             }
         }
+        else if (ActionBitIsSet(action, KAUTH_VNODE_DELETE))
+        {
+            if (!TrySendRequestAndWaitForResponse(
+                    root,
+                    MessageType_KtoU_NotifyDirectoryPreDelete,
+                    currentVnode,
+                    pid,
+                    procname,
+                    &kauthResult,
+                    kauthError))
+            {
+                goto CleanupAndReturn;
+            }
+        }
     }
     else
     {
-        if (ActionBitIsSet(
+        if (ActionBitIsSet(action, KAUTH_VNODE_DELETE))
+        {
+            if (!TrySendRequestAndWaitForResponse(
```

Comment:
Though I'm curious, why are you triggering pre-delete before hydration? Does the ordering make any difference?

Reply:
> Does the ordering make any difference?

If the provider decides to block the delete it avoids an unnecessary hydration (I don't think VFSForGit would ever do that though ...)

Assuming the provider *allows* the delete, I don't think the ordering would make a difference.  You could argue that the provider wouldn't get the PreDelete notification if the hydration were to fail, but I'm not sure that's enough of a reason to re-order these.

Reply:
See related comment, I'll refactor this code to avoid duplication (and to hydrate first)

Reply:
Update: For now I've removed all of the "hydrate on delete" code.  Interestingly, it seems like we still get some kauth request that hydrates the file before moving.  I'll test more in this area when adding support for rename notifications.